File: python/sequence.py
# -*- coding: utf-8 -*-
u"""
Возьмём бесконечную цифровую последовательность, образованную склеиванием 
последовательных положительных чисел: S = 123456789101112131415...
Определите первое вхождение заданной последовательности A в бесконечной 
последовательности S (нумерация начинается с 1).
"""

def get_subsequence_pos(subsequence):
	subsequence = str(subsequence)
	subsequence_len = len(subsequence)
	sequence, sequence_len, sequence_counter = '', 0, 0

	while sequence != subsequence:
		sequence_counter += 1
		str_sequence_counter = str(sequence_counter)
		for c in str_sequence_counter:
			sequence_len += 1

			if sequence_len > subsequence_len:
				sequence = sequence[1:] + c  # чтобы не хранить всю последовательность в памяти
			else:
				sequence += c

			if sequence == subsequence:
				break;

	return sequence_len - subsequence_len + 1

print get_subsequence_pos(6789)  # => 6
print get_subsequence_pos(101)  # => 10

File: java/Sequence/src/Sequence.java
/*
Возьмём бесконечную цифровую последовательность, образованную склеиванием 
последовательных положительных чисел: S = 123456789101112131415...
Определите первое вхождение заданной последовательности A в бесконечной 
последовательности S (нумерация начинается с 1).
*/

public class Sequence {
	public static void main (String[] args) {
		Service serv = new Service();

		System.out.println(serv.getSubsequencePos("6789"));
		System.out.println(serv.getSubsequencePos("101"));
	}	
}

class Service {
	public int getSubsequencePos(String subsequence) {
		int subsequenceLen = subsequence.length();
		String sequence = "";
		int sequenceLen = 0, sequenceCounter = 0;
		

		while(!sequence.equals(subsequence)) {
			sequenceCounter++;
			String strSequenceCounter = sequenceCounter + "";

			for(int i = 0; i < strSequenceCounter.length(); i++) {
				char c = strSequenceCounter.charAt(i);
				sequenceLen++;

				if(sequenceLen > subsequenceLen) {
					sequence = sequence.substring(1) + c; // чтобы не хранить всю последовательность в памяти
				}
				else {
					sequence += c;
				}

				if(sequence.equals(subsequence)) {
					break;
				}
			}
		}

		return sequenceLen - subsequenceLen + 1;
	}
}